lib/rules/callback-return.js

```diff
@@ -0,0 +1,136 @@
+ * @author Jamund Ferguson
+ * @copyright 2015 Jamund Ferguson. All rights reserved.
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+
+    var callbacks = context.options[0] || ["callback", "cb", "next"];
+
+    //--------------------------------------------------------------------------
+    // Helpers
+    //--------------------------------------------------------------------------
+
+    /**
+     * Find the closest parent matching a list of types.
+     * @param {ASTNode} node The node whose parents we are searching
+     * @param {Array} types The node types to match
+     * @returns {ASTNode} The matched node or undefined.
+     */
+    function findClosestParentOfType(node, types) {
+        if (!node.parent) {
+            return null;
+        }
+        if (types.indexOf(node.parent.type) === -1) {
+            return findClosestParentOfType(node.parent, types);
+        }
+        return node.parent;
+    }
+
+    /**
+     * Check to see if a CallExpression is in our callback list.
+     * @param {ASTNode} node The node to check against our callback names list.
+     * @returns {Boolean} Whether or not this function matches our callback name.
+     */
+    function isCallback(node) {
+        return node.callee.type === "Identifier" && callbacks.indexOf(node.callee.name) > -1;
+    }
+
+    /**
+     * Determines whether or not the callback is part of a callback expression.
+     * @param {ASTNode} node The callback node
+     * @param {ASTNode} parentNode The expression node
+     * @returns {boolean} Whether or not this is part of a callback expression
+     */
+    function isCallbackExpression(node, parentNode) {
+
+        // ensure the parent node exists and is an expression
+        if (!parentNode || parentNode.type !== "ExpressionStatement") {
+            return false;
+        }
+
+        // cb()
+        if (parentNode.expression === node) {
+            return true;
+        }
+
+        // special case for cb && cb() and similar
+        if (parentNode.expression.type === "BinaryExpression" || parentNode.expression.type === "LogicalExpression") {
+            if (parentNode.expression.right === node) {
+                return true;
+            }
+        }
+    }
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    return {
+        "CallExpression": function(node) {
+
+            // if we"re not a callback we can return
+            if (!isCallback(node)) {
+                return;
+            }
+
+            // find the closest block, return or if statement
+            var closestBlock = findClosestParentOfType(node, ["BlockStatement", "ReturnStatement", "IfStatement"]),
```

Comment:
You specifically talking about line 107? https://github.com/eslint/eslint/pull/2828/files#diff-c53c86e3156c85a799267c9ec841d054R107